fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java

```diff
@@ -1442,6 +1452,70 @@ public static TWorkGroup prepareResourceGroup(ConnectContext connect, ResourceGr
         return resourceGroup;
     }
 
+    private List<Pair<Long, TNetworkAddress>> adaptiveChooseNodes(PlanFragment fragment, Map<Long, ComputeNode> candidates,
+                                                                  Set<TNetworkAddress> childUsedHosts) {
+        List<Pair<Long, TNetworkAddress>> childHosts = Lists.newArrayList();
+        for (Map.Entry<Long, ComputeNode> entry : candidates.entrySet()) {
+            TNetworkAddress address = entry.getValue().getAddress();
+            if (childUsedHosts.contains(address)) {
+                childHosts.add(Pair.create(entry.getKey(), address));
+            }
+        }
+
+        // sometimes we may reserve the right fragment and add the children of left fragment
+        // in the children list of right fragment like SHUFFLE_HASH_BUCKET plan, so we need sort
+        // the list to ensure the most left child is at the 0 index position.
+        List<PlanFragment> sortedFragments = fragment.getChildren().stream()
+                .sorted(Comparator.comparing(e -> e.getPlanRoot().getId().asInt()))
+                .collect(Collectors.toList());
+
+        long maxOutputOfRightChild = sortedFragments.stream().skip(1)
+                .map(e -> e.getPlanRoot().getCardinality()).reduce(Long::max)
+                .orElse(fragment.getChild(0).getPlanRoot().getCardinality());
+        long outputOfMostLeftChild = sortedFragments.get(0).getPlanRoot().getCardinality();
+
+        long baseNodeNums = Math.max(1, maxOutputOfRightChild / TINY_SCALE_ROWS_LIMIT / fragment.getPipelineDop());
```

Comment:
can `fragment.getPipelineDop()` be 0?

Reply:
We had set the real dop  before these step, so it cannot be 0.